src/core/thread/mle.cpp

```diff
@@ -1333,35 +1333,58 @@ bool Mle::HasUnregisteredAddress(void)
 
 otError Mle::AppendAddressRegistration(Message &aMessage, AddressRegistrationMode aMode)
 {
-    otError                  error = OT_ERROR_NONE;
-    Tlv                      tlv;
-    AddressRegistrationEntry entry;
-    Lowpan::Context          context;
-    bool                     done        = false;
-    uint8_t                  length      = 0;
-    uint8_t                  counter     = 0;
-    uint16_t                 startOffset = aMessage.GetLength();
+    otError                         error = OT_ERROR_NONE;
+    Tlv                             tlv;
+    AddressRegistrationEntry        entry;
+    Lowpan::Context                 context;
+    bool                            meshLocalOnlyDone = false;
+    bool                            meshLocalDone     = false;
+    uint8_t                         length            = 0;
+    uint8_t                         counter           = 0;
+    uint16_t                        startOffset       = aMessage.GetLength();
+    const Ip6::NetifUnicastAddress *next              = NULL;
 
     tlv.SetType(Tlv::kAddressRegistration);
     SuccessOrExit(error = aMessage.Append(&tlv, sizeof(tlv)));
 
-    for (const Ip6::NetifUnicastAddress *addr = Get<ThreadNetif>().GetUnicastAddresses(); addr; addr = addr->GetNext())
+    for (const Ip6::NetifUnicastAddress *addr = Get<ThreadNetif>().GetUnicastAddresses(); addr; addr = next)
```

Comment:
How about before the `for` loop over all unicast addresses, we add an entry for `GetMeshLocal64()` address to the message and then go through the loop of all address (now skipping over the address matching ML64 ).

 I think this will also help simplify the case `mMode == kAppendMeshLocalOnly`.

Thoughts?

Reply:
Make sense. Updated. Thanks.